File: packages/typia/src/schemas/metadata/MetadataNative.ts
import { ClassProperties } from "../../typings/ClassProperties";

import { IMetadataNative } from "./IMetadataNative";
import { IMetadataTypeTag } from "./IMetadataTypeTag";

export class MetadataNative {
  public readonly name: string;
  public readonly tags: IMetadataTypeTag[][];
  private typeName_?: string;

  private constructor(props: ClassProperties<MetadataNative>) {
    this.name = props.name;
    this.tags = props.tags;
  }

  public static create(props: ClassProperties<MetadataNative>): MetadataNative {
    return new MetadataNative(props);
  }

  public getName(): string {
    return (this.typeName_ ??= (() => {
      if (this.tags.length === 0) return this.name;
      else if (this.tags.length === 1) {
        const str: string = [
          this.name,
          ...this.tags[0]!.map((t) => t.name),
        ].join(" & ");
        return `(${str})`;
      }
      const rows: string[] = this.tags.map((row) => {
        const str: string = row.map((t) => t.name).join(" & ");
        return row.length === 1 ? str : `(${str})`;
      });
      return `(${this.name} & (${rows.join(" | ")}))`;
    })());
  }

  public toJSON(): IMetadataNative {
    return {
      name: this.name,
      tags: this.tags,
    };
  }
}

File: packages/core/src/schemas/metadata/MetadataNative.go
package metadata

type MetadataNative struct {
	Name string
	Tags [][]MetadataTypeTag

	typeName string
}

type MetadataNativeProps struct {
	Name string
	Tags [][]MetadataTypeTag
}

func CreateMetadataNative(props MetadataNativeProps) *MetadataNative {
	return &MetadataNative{
		Name: props.Name,
		Tags: cloneTagMatrix(props.Tags),
	}
}

func (m *MetadataNative) GetName() string {
	if m.typeName == "" {
		m.typeName = taggedReferenceName(m.Name, m.Tags)
	}
	return m.typeName
}

func (m *MetadataNative) ToJSON() MetadataReferenceJSON {
	return MetadataReferenceJSON{
		Name: m.Name,
		Tags: cloneTagMatrix(m.Tags),
	}
}
